<a href="https://colab.research.google.com/github/mehjabeenalam-ux/REC-2-Machine-Learning/blob/main/Mini_Z_Score_Project_PandasEDA.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

How to Calculate Z-Score in Pandas (Practical Code)

Import Pandas Library

In [3]:
import pandas as pd

Sample Employee Dataset

In [24]:
df = pd.DataFrame({
    'name': ['Alice', 'Bob', 'Charlie', 'David', 'Eve'],
    'salary': [70000, 120000, 85000, 95000, 45000]
})

print(f"Employee Dataset {df}")

Employee Dataset       name  salary
0    Alice   70000
1      Bob  120000
2  Charlie   85000
3    David   95000
4      Eve   45000


Shape and Size of the Employee Dataset

In [11]:
print(df.shape)
print(df.size) #Row*Column

(5, 3)
15


Summary Statistics of the Employee Dataset

In [9]:
df.describe().round((2))

salary  z_salary
count       5.0      5.00
mean    83000.0      0.00
std     27973.2      1.00
min     45000.0     -1.36
25%     70000.0     -0.46
50%     85000.0      0.07
75%     95000.0      0.43
max    120000.0      1.32

Calculate Z-score for salary column

In [7]:
df['z_salary'] = (df['salary'] - df['salary'].mean()) / df['salary'].std()

print(df[['name', 'salary', 'z_salary']].round(2))

      name  salary  z_salary
0    Alice   70000     -0.46
1      Bob  120000      1.32
2  Charlie   85000      0.07
3    David   95000      0.43
4      Eve   45000     -1.36


Sorting the Z Score of the Employee Dataset

Sort Z-scores from lowest to highest (smallest → largest)

In [23]:
sorted_low_to_high = df.sort_values('z_salary')
print("Sorted Z-scores (lowest to highest):")
print(sorted_low_to_high[['name', 'salary', 'z_salary']].round(3))

Sorted Z-scores (lowest to highest):
      name  salary  z_salary
4      Eve   45000    -1.358
0    Alice   70000    -0.465
2  Charlie   85000     0.071
3    David   95000     0.429
1      Bob  120000     1.323


Interpretation of the above Z Score

- Bob has z_salary = 1.32 → salary is 1.32 standard deviations above the mean (high earner, possible outlier on the high side)

- Eve has z_salary = -1.36 → salary is 1.36 standard deviations below the mean (low earner)

- Charlie is very close to 0 → salary is near the average

Rule of thumb:

- |Z| > 3 → strong outlier; |Z| > 2 → moderate outlier (common thresholds)

Lowest Salary Outlier

In [19]:
outlier_salary1 = round(df['z_salary'].min(), 2)
print("Lowest Salary Outlier to the mean (Z-score nearest 0):\n", outlier_salary1)

Lowest Salary Outlier to the mean (Z-score nearest 0):
 -1.36


Highest Salary Outlier

In [18]:
outlier_salary = round(df['z_salary'].max(), 2)
print("Highest Salary Outlier to the mean (Z-score nearest 0):\n", outlier_salary)

Highest Salary Outlier to the mean (Z-score nearest 0):
 1.32


Row with Z-score closest to the mean

In [15]:
closest_to_mean = df.loc[df['z_salary'].abs() == df['z_salary'].abs().min()]
print("Row closest to the mean (Z-score nearest 0):\n", closest_to_mean)

Row closest to the mean (Z-score nearest 0):
       name  salary  z_salary
2  Charlie   85000  0.071497
